### PR TITLE
feat[marketplace]: add slot queue pausing

### LINKED ISSUE #549: Introduce Slot Queue pausing and remove past events fetching
When the Slot Queue runs out of availabilities to fill, it should pause until new availability is added or some capacity is returned to existing availabilities.

Additionally, when new availability is added, the fetching of past events in the availability-added callback should be removed. This is because the queue will be paused when no availability is present, and slots will only be removed when they are filled or when there is availability available but non-matching.

### PR DESCRIPTION
**Pull request by emizzle**
Adds slot queue pausing based on the design in https://github.com/codex-storage/codex-research/pull/188.

@AuHau -- I took the changes in commit https://github.com/codex-storage/nim-codex/pull/748/commits/1425bf5c3f42a80f16821789af7957d4757f1e17 and added to this PR as I needed to update that routine.

Close #549

### CODE REVIEW

File: codex/sales/slotqueue.nim
Comment (AuHau) [inline codex/sales/slotqueue.nim:271]: While I guess it can be fired everytime, shouldn't it be intended under the `if` above?
```suggestion
    self.unpause()
```

File: codex/sales/slotqueue.nim
Comment (AuHau) [inline codex/sales/slotqueue.nim:363]: Not finished comment ;-)

> first pop all items

Hmm I don't think this is what is happening bellow right? `self[i]` only access `i-th` element and not "pop" it... Right?

**Review by AuHau — COMMENTED** [review]
Nice work! 👍

However, I am a bit confused about one part of your PR. In your proposal at https://github.com/codex-storage/codex-research/pull/188, there were no mentions of the concept of "empty availabilities". I remember we discussed this before and I think reaching the definition of what "empty availability" is was bit tough.

I noticed that you have two ways to pause the queue. The first is based on the `seen` flag as per your original proposal. The second is triggered when "all availabilities are empty". I am unsure why you added the second part. Wouldn't the `seen` flag mechanism already cover this scenario? It seems to introduce unnecessary complexity and complicates understanding and reasoning about the codebase.

What do you think?

File: codex/sales/slotqueue.nim
Comment (emizzle) [reply codex/sales/slotqueue.nim:363]: Yep i figured out a better way and forgot to update the comment. Will update.

File: codex/sales/slotqueue.nim
Comment (emizzle) [reply codex/sales/slotqueue.nim:271]: Yep, it doesn't matter but can update for optics.

File: codex/sales/reservations.nim
Comment (emizzle) [reply codex/sales/reservations.nim:98]: It's a forward declaration. Allows calls to the routine in the module when the routine is defined further down the module.

File: codex/sales/states/errored.nim
Comment (emizzle) [reply codex/sales/states/errored.nim:33]: A could have happened in any of the states, eg downloading.

**Comment by emizzle** [comment]
That's a great point -- the queue passing should work without the added complexity of pausing when availabilities are "empty". I'll try to remove and write a test to assert that the queue pauses once availabilities are "emptied".

File: codex/sales/states/errored.nim
Comment (emizzle) [reply codex/sales/states/errored.nim:33]: Thinking about this again, if an error occurred in preparing, downloading, or filling, then it stands to reason that the node should take another stab at filling the slot. If an occurred in filled or proving, then the node should not try to reprocess the slot. Will update to reflect this.

**Comment by emizzle** [comment]
@AuHau the PR has been updated with your feedback. Thanks for the review! 🙌

File: codex/sales/states/errored.nim
Comment (emizzle) [reply codex/sales/states/errored.nim:33]: I updated it so that only when `SaleErrored` is reached from `SaleDownloading` will the slot be reprocessed. `SalePreparing`. I was thinking that perhaps we could extend this same behaviour to SalePreparing and SaleInitialProving, however I'm not entirely sure if we would want to reprocess a slot if there's an exception while preparing (availability checking and reservation creation), or initial proving (generating the proof).

File: tests/codex/sales/testsales.nim
Comment (AuHau) [inline tests/codex/sales/testsales.nim:295]: Dead code? Can be deleted?

File: tests/codex/sales/testsales.nim
Comment (AuHau) [inline tests/codex/sales/testsales.nim:277]: Another dead code?

File: codex/sales/states/errored.nim
Comment (AuHau) [reply codex/sales/states/errored.nim:33]: Nice!
Yeah, lets discuss this on call today. I would maybe incline toward reprocessing for Preparing but not InitialProving?

**Review by AuHau — APPROVED** [review]
Pretty great work! There are some cleanup items, but generally LGTM. 

Before merging lets discuss on our call about the reprocessing of error states.

**Comment by emizzle** [comment]
This PR should be ready to go, however I've some (small) changes to the way the queue pauses, so maybe it might be prudent to review again @AuHau ?

**Review by AuHau — APPROVED** [review]
Just few comments but generally LGTM! Great work! 👍

File: codex/sales/slotqueue.nim
Comment (markspanbroek) [inline codex/sales/slotqueue.nim:225]: It's the other way around; `clear()` ensures that `wait()`s are blocked, `fire()` ensures that `wait()`s go through. The documentation of `AsyncEvent` is not very good on this point, but the description of `fire()` is clear: https://status-im.github.io/nim-chronos/api/chronos/asyncsync.html#fire%2CAsyncEvent

```suggestion
proc pause*(self: SlotQueue) =
  # set unpaused flag to false -- coroutines will block on unpaused.wait()
  self.unpaused.clear()

proc unpause*(self: SlotQueue) =
  # set unpaused flag to true -- unblocks coroutines waiting on unpaused.wait()
  self.unpaused.fire()

```

File: codex/sales/slotqueue.nim
Comment (markspanbroek) [reply codex/sales/slotqueue.nim:379]: `update()` only returns false when the argument is not in the queue. Since we're supplying `self.queue[0]` as the arguments, this will never happen. And if it does happen, then it's a programmer error and the node should crash, I think.

File: codex/sales/slotqueue.nim
Comment (markspanbroek) [inline codex/sales/slotqueue.nim:405]:
```suggestion
      # block until unpaused is true/fired, ie wait for queue to be unpaused
```

File: codex/sales/states/downloading.nim
Comment (markspanbroek) [inline codex/sales/states/downloading.nim:34]:
```suggestion
# override onError so that the slot is reprocessed in the case of an unhandled
```

File: codex/sales/states/downloading.nim
Comment (markspanbroek) [inline codex/sales/states/downloading.nim:37]: I know we discussed this before, but I can't remember the conclusion. Did we want to try a slot again if it failed to download the first time? I would think that we don't want to do this, because there's no reason to assume that it would work a second time around, so it's better to focus on other slots?

File: codex/sales/states/downloading.nim
Comment (markspanbroek) [inline codex/sales/states/downloading.nim:77]: Same as above

**Review by markspanbroek — DISMISSED** [review]
Great work @emizzle, thanks! I have a few small comments, but other than that good to go!

**Comment by dryajov** [comment]
Where are we with this? Should it be merged?

File: codex/sales/slotqueue.nim
Comment (emizzle) [reply codex/sales/slotqueue.nim:379]: Yes, exactly. In addition, we know that self.queue[0] exists because there is an early return
when `if self.queue.empty` is true. If the queue is empty when we don't expect
it to be, there is no real way to recover.

File: codex/sales.nim
Comment (emizzle) [reply codex/sales.nim:267]: Removed and updated the comments to reflect.

File: codex/sales/slotqueue.nim
Comment (emizzle) [reply codex/sales/slotqueue.nim:225]: Thanks for the clarification. I had forgotten to update the comments, but got it
right in the logic 😅

File: codex/sales/states/downloading.nim
Comment (emizzle) [reply codex/sales/states/downloading.nim:37]: Yes, that's correct. We did discuss it and I have it in my notes that it should
be updated so that errors are not reprocessed. But it was so long ago, I forgot
to do it. Great catch, thank you!

File: codex/sales/states/downloading.nim
Comment (emizzle) [reply codex/sales/states/downloading.nim:34]: Good catch, will remove the entire override

**Comment by emizzle** [comment]
> Where are we with this? Should it be merged?

Will merge once green. Had to address a few small comments and rebase on master.

**Review by markspanbroek — APPROVED** [review]

File: codex/sales/slotqueue.nim
Comment (emizzle) [reply codex/sales/slotqueue.nim]: I know where I got this mixed up. Look at the [documentation for the constructor for `AsyncEvent`](https://status-im.github.io/nim-chronos/api/chronos/asyncsync.html#AsyncEvent): 
> An event manages a flag that can be set to true with the fire() procedure and reset to false with the clear() procedure. _The wait() coroutine blocks until the flag is false._ (emphasis mine)

This is actually wrong, or at least I seemed to keep reading it in way that says the flag must be false to unblock coroutines.

**Review by AuHau — APPROVED** [review]

**Pull request merged**
